certora/specs/Reverts.spec

```diff
@@ -0,0 +1,183 @@
+    requireInvariant preLIFConsistent();
+
+    preLiquidate@withrevert(e, borrower, seizedAssets, repaidShares, data);
+    assert !summaryExactlyOneZero(seizedAssets, repaidShares) => lastReverted;
+}
+
+// Check that collateralQuoted == 0 would revert by failing require-statements.
+rule zeroCollateralQuotedReverts(env e, address borrower, uint256 seizedAssets, bytes data) {
+    requireInvariant preLltvConsistent();
+
+    uint256 collateralQuoted;
+    uint256 borrowed;
+
+    uint256 higherBound = summaryWMulDown(collateralQuoted, currentContract.LLTV);
+    uint256 lowerBound = summaryWMulDown(collateralQuoted, currentContract.PRE_LLTV);
+
+    assert  collateralQuoted == 0 => (lowerBound >= borrowed || borrowed > higherBound);
+}
+
+// Check that pre-liquidating a position such that LTV <= PRE_LLTV reverts.
+// This also implies that LTV > PRE_LLTV when borrowed > collateralQuoted.summaryWMulDown(PRE_LLTV).
+rule nonLiquidatablePositionReverts(env e, address borrower, uint256 seizedAssets, bytes data) {
+    requireInvariant preLltvConsistent();
+    requireInvariant preLCFConsistent();
+    requireInvariant preLIFConsistent();
+
+    uint256 mLastUpdate;
+    (_, _, _, _, mLastUpdate, _) = MORPHO.market(currentContract.ID);
+
+    // Ensure that no interest is accumulated.
+    require mLastUpdate == e.block.timestamp;
+
+    mathint ltv = getLtv(borrower);
+
+    preLiquidate@withrevert(e, borrower, seizedAssets, 0, data);
+
+    // Ensure the price is unchanged in the preLiquidate call.
+    require !priceChanged;
+
+    assert ltv <= currentContract.PRE_LLTV => lastReverted;
+}
+
+// Check that pre-liquidating a position such that LTV > LLTV would revert.
+rule liquidatablePositionReverts(env e, address borrower, uint256 seizedAssets, bytes data) {
+    requireInvariant preLltvConsistent();
+    requireInvariant preLCFConsistent();
+    requireInvariant preLIFConsistent();
+
+    uint256 mLastUpdate;
+    (_, _, _, _, mLastUpdate, _) = MORPHO.market(currentContract.ID);
+
+    // Ensure that no interest is accumulated.
+    require mLastUpdate == e.block.timestamp;
+
+    mathint ltv = getLtv(borrower);
+
+    preLiquidate@withrevert(e, borrower, seizedAssets, 0, data);
+
+    // Ensure the price is unchanged in the preLiquidate call.
+    require !priceChanged;
+
+    assert ltv > currentContract.LLTV => lastReverted;
+}
+
+// Check that a pre-liquidation that repays more shares than available or allowed by the preLCF reverts.
+rule excessivePreliquidationWithAssetsReverts(env e, address borrower, uint256 seizedAssets, bytes data) {
+    // Market values.
+    uint256 mTotalBorrowAssets;
+    uint256 mTotalBorrowShares;
+    uint256 mLastUpdate;
+
+    // Position values.
+    uint256 pBorrowShares;
+
+    requireInvariant preLltvConsistent();
+    requireInvariant preLCFConsistent();
+    requireInvariant preLIFConsistent();
+
+    (_, _, mTotalBorrowAssets, mTotalBorrowShares, mLastUpdate, _) = MORPHO.market(currentContract.ID);
```

Comment:
Note that we could use a `market_` getter that would return a struct, which would make it look much nicer IMO. See [this issue](https://github.com/morpho-org/morpho-blue/issues/699)

Reply:
Implemented in https://github.com/morpho-org/pre-liquidation/pull/37/commits/350020475fe843e54d61676820b7ce55ecd6eff2.

Interestingly, I attempted to do something similar for `position` but somehow, I didn't call managed to fix the call resolution see this [run](https://prover.certora.com/output/129828/c00703fdd015413a8321bdea5f25202c/?anonymousKey=534fcdc38ef2f819811fd6661e5af12c1276e3be).

Reply:
This should really be implemented at the Morpho Blue level, as suggested in the issue (notice that this is an issue on morpho-blue repository). Otherwise this makes an extra CALL which is not wanted, and may cause havocs that we don't want to have to deal with

Reply:
Sorry I'm confused, should I revert to the previous implementation?

Reply:
Oh right, that is clearer indeed, thank you 😄 .

Reply:
Implemented here https://github.com/morpho-org/pre-liquidation/pull/37/commits/8c4110a8134c977d27abc287859c44e362cde28e.